File: apps/api/momento_api/routes/tags.py
from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException, status

from momento_api.auth.dependencies import CurrentUser, get_current_user, require_admin
from momento_api.database import execute_query, fetch_all, fetch_one, get_connection, insert_returning_id
from momento_api.models.tag import (
    TagAddToMediaRequest,
    TagCreateRequest,
    TagDeleteRequest,
    TagListResponse,
    TagRemoveFromMediaRequest,
    TagResponse,
)

router = APIRouter(prefix="/tag", tags=["tag"])


def _row_to_tag_response(row) -> TagResponse:
    return TagResponse(id=row["id"], name=row["name"], created_at=row["created_at"])


@router.post("/list", response_model=TagListResponse)
async def list_tags(_: Annotated[CurrentUser, Depends(get_current_user)]) -> TagListResponse:
    rows = fetch_all("SELECT * FROM tags ORDER BY name", ())
    return TagListResponse(tags=[_row_to_tag_response(row) for row in rows])


@router.post("/create", response_model=TagResponse)
async def create_tag(request: TagCreateRequest, _: Annotated[CurrentUser, Depends(get_current_user)]) -> TagResponse:
    existing = fetch_one("SELECT id FROM tags WHERE name = ?", (request.name,))
    if existing:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Tag already exists")

    tag_id = insert_returning_id("INSERT INTO tags (name) VALUES (?)", (request.name,))

    row = fetch_one("SELECT * FROM tags WHERE id = ?", (tag_id,))
    if row is None:
        raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail="Failed to create tag")

    return _row_to_tag_response(row)


@router.post("/delete")
async def delete_tag(request: TagDeleteRequest, _: Annotated[CurrentUser, Depends(require_admin)]) -> dict:
    tag_row = fetch_one("SELECT id FROM tags WHERE id = ?", (request.tag_id,))
    if tag_row is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Tag not found")

    execute_query("DELETE FROM tags WHERE id = ?", (request.tag_id,))
    get_connection().commit()

    return {"message": "Tag deleted successfully"}


@router.post("/add-to-media")
async def add_tag_to_media(
    request: TagAddToMediaRequest, current_user: Annotated[CurrentUser, Depends(get_current_user)]
) -> dict:
    tag_row = fetch_one("SELECT id FROM tags WHERE id = ?", (request.tag_id,))
    if tag_row is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Tag not found")

    conn = get_connection()
    params: list[tuple[int, int]] = []
    for media_id in request.media_ids:
        media_row = fetch_one("SELECT id FROM media WHERE id = ? AND user_id = ?", (media_id, current_user.id))
        if media_row is None:
            continue
        params.append((media_id, request.tag_id))

    if not params:
        return {"message": "Tag added to media"}

    conn.executemany("INSERT OR IGNORE INTO media_tags (media_id, tag_id) VALUES (?, ?)", params)
    conn.commit()

    return {"message": "Tag added to media"}


@router.post("/remove-from-media")
async def remove_tag_from_media(
    request: TagRemoveFromMediaRequest, current_user: Annotated[CurrentUser, Depends(get_current_user)]
) -> dict:
    conn = get_connection()
    for media_id in request.media_ids:
        media_row = fetch_one("SELECT id FROM media WHERE id = ? AND user_id = ?", (media_id, current_user.id))
        if media_row:
            conn.execute("DELETE FROM media_tags WHERE media_id = ? AND tag_id = ?", (media_id, request.tag_id))
    conn.commit()

    return {"message": "Tag removed from media"}

File: apps/api/src/routes/tags.rs
use axum::{extract::State, routing::post, Json, Router};

use crate::auth::{AppState, CurrentUser, RequireAdmin};
use crate::database::{execute_query, fetch_all, fetch_one, insert_returning_id, queries};
use crate::error::{AppError, AppResult};
use crate::models::{
    TagAddToMediaRequest, TagCreateRequest, TagDeleteRequest, TagListResponse,
    TagRemoveFromMediaRequest, TagResponse,
};

pub fn router() -> Router<AppState> {
    Router::new()
        .route("/tag/list", post(list_tags))
        .route("/tag/create", post(create_tag))
        .route("/tag/delete", post(delete_tag))
        .route("/tag/add-to-media", post(add_tag_to_media))
        .route("/tag/remove-from-media", post(remove_tag_from_media))
}

fn map_tag_row(row: &rusqlite::Row) -> rusqlite::Result<TagResponse> {
    Ok(TagResponse {
        id: row.get(0)?,
        name: row.get(1)?,
        created_at: row.get(2)?,
    })
}

async fn list_tags(
    State(state): State<AppState>,
    _current_user: CurrentUser,
) -> AppResult<Json<TagListResponse>> {
    let conn = state.pool.get().map_err(AppError::Pool)?;

    let tags = fetch_all(&conn, queries::tags::SELECT_ALL, &[], map_tag_row)?;

    Ok(Json(TagListResponse { tags }))
}

async fn create_tag(
    State(state): State<AppState>,
    _current_user: CurrentUser,
    Json(request): Json<TagCreateRequest>,
) -> AppResult<Json<TagResponse>> {
    let conn = state.pool.get().map_err(AppError::Pool)?;

    // Check existing
    let existing = fetch_one(
        &conn,
        queries::tags::SELECT_ID_BY_NAME,
        &[&request.name],
        |row| row.get::<_, i64>(0),
    )?;

    if existing.is_some() {
        return Err(AppError::BadRequest("Tag already exists".to_string()));
    }

    let tag_id = insert_returning_id(&conn, queries::tags::INSERT, &[&request.name])?;

    let tag = fetch_one(&conn, queries::tags::SELECT_BY_ID, &[&tag_id], map_tag_row)?
        .ok_or_else(|| AppError::Internal("Failed to create tag".to_string()))?;

    Ok(Json(tag))
}

async fn delete_tag(
    State(state): State<AppState>,
    RequireAdmin(_): RequireAdmin,
    Json(request): Json<TagDeleteRequest>,
) -> AppResult<Json<serde_json::Value>> {
    let conn = state.pool.get().map_err(AppError::Pool)?;

    let exists = fetch_one(
        &conn,
        queries::tags::CHECK_EXISTS,
        &[&request.tag_id],
        |row| row.get::<_, i64>(0),
    )?;

    if exists.is_none() {
        return Err(AppError::NotFound("Tag not found".to_string()));
    }

    execute_query(&conn, queries::tags::DELETE, &[&request.tag_id])?;

    Ok(Json(
        serde_json::json!({"message": "Tag deleted successfully"}),
    ))
}

async fn add_tag_to_media(
    State(state): State<AppState>,
    current_user: CurrentUser,
    Json(request): Json<TagAddToMediaRequest>,
) -> AppResult<Json<serde_json::Value>> {
    let conn = state.pool.get().map_err(AppError::Pool)?;

    let tag_exists = fetch_one(
        &conn,
        queries::tags::CHECK_EXISTS,
        &[&request.tag_id],
        |row| row.get::<_, i64>(0),
    )?;

    if tag_exists.is_none() {
        return Err(AppError::NotFound("Tag not found".to_string()));
    }

    for media_id in &request.media_ids {
        // Check media belongs to user
        let media_exists = fetch_one(
            &conn,
            queries::media::CHECK_EXISTS,
            &[media_id, &current_user.id],
            |row| row.get::<_, i64>(0),
        )?;

        if media_exists.is_none() {
            continue;
        }

        let _ = conn.execute(
            queries::tags::ADD_TO_MEDIA,
            rusqlite::params![media_id, request.tag_id],
        );
    }

    Ok(Json(serde_json::json!({"message": "Tag added to media"})))
}

async fn remove_tag_from_media(
    State(state): State<AppState>,
    current_user: CurrentUser,
    Json(request): Json<TagRemoveFromMediaRequest>,
) -> AppResult<Json<serde_json::Value>> {
    let conn = state.pool.get().map_err(AppError::Pool)?;

    for media_id in &request.media_ids {
        // Check media belongs to user
        let media_exists = fetch_one(
            &conn,
            queries::media::CHECK_EXISTS,
            &[media_id, &current_user.id],
            |row| row.get::<_, i64>(0),
        )?;

        if media_exists.is_some() {
            conn.execute(
                queries::tags::REMOVE_FROM_MEDIA,
                rusqlite::params![media_id, request.tag_id],
            )?;
        }
    }

    Ok(Json(
        serde_json::json!({"message": "Tag removed from media"}),
    ))
}
